web/src/changehash.js

```diff
@@ -0,0 +1,72 @@
+import * as blueslip from "./blueslip";
+import * as browser_history from "./browser_history";
+import * as hash_util from "./hash_util";
+import * as message_viewport from "./message_viewport";
+import {user_settings} from "./user_settings";
+
+export function get_full_url(hash) {
+    const location = window.location;
+
+    if (hash.charAt(0) !== "#" && hash !== "") {
+        hash = "#" + hash;
+    }
+
+    // IE returns pathname as undefined and missing the leading /
+    let pathname = location.pathname;
+    if (pathname === undefined) {
+        pathname = "/";
+    } else if (pathname === "" || pathname.charAt(0) !== "/") {
+        pathname = "/" + pathname;
+    }
+
+    // Build a full URL to not have same origin problems
+    const url = location.protocol + "//" + location.host + pathname + hash;
+    return url;
+}
+
+export function set_hash(hash) {
+    if (hash === window.location.hash) {
+        // Avoid adding duplicate entries in browser history.
+        return;
+    }
+    if (history.pushState) {
+        const url = get_full_url(hash);
+        try {
+            history.pushState(null, null, url);
+            browser_history.update_web_public_hash(hash);
+        } catch (error) {
+            if (error instanceof TypeError) {
+                // The window has been destroyed and the history object has been marked dead, so cannot
+                // be updated.  Silently do nothing, since there's nothing we can do.
+            } else {
+                throw error;
+            }
+        }
+    } else {
+        // pushState has 97% global support according to caniuse. So, we will ideally never reach here.
+        // TODO: Delete this case if we don't see any error reports in a while.
+        if (hash === "" || hash === "#") {
+            // Setting empty hash here would scroll to the top.
+            hash = user_settings.default_view;
+        }
+
+        blueslip.error("browser does not support pushState");
+        window.location.hash = hash;
+    }
+}
```

Comment:
I think at least `set_hash` at least should live in `browser_history.ts`; `browser_history.set_hash` seems like a plausible interface?

Reply:
Sure. Note that this demotes `browser_history` from full Node test coverage status (but that’s a weird module to have Node tests anyway, since those tell you nothing about what will happen in a browser).